index/scorch/segment/zap/merge.go

```diff
@@ -361,14 +349,29 @@ func persistMergedRest(segments []*SegmentBase, dropsIn []*roaring.Bitmap,
 
 		// update the field doc values
 		fdvEncoder := newChunkedContentCoder(uint64(chunkFactor), newSegDocCount-1, w, true)
-		for docNum, docTerms := range docTermMap {
-			if len(docTerms) > 0 {
-				err = fdvEncoder.Add(uint64(docNum), docTerms)
-				if err != nil {
-					return nil, 0, err
+
+		newDocNum := uint64(0)
+		for segmentI, segment := range segmentsInFocus {
+			fdvReader := segment.fieldDvReaders[uint16(fieldID)]
+			if fdvReader == nil {
+				continue
+			}
+
+			dropI := drops[segmentI]
+
+			err = fdvReader.iterateDocTerms(segment, func(docNum uint64, terms []byte) error {
+				if dropI != nil && dropI.Contains(uint32(docNum)) {
+					return nil
 				}
+				err := fdvEncoder.Add(newDocNum, terms)
+				newDocNum++
```

Comment:
hmmm, this newDocNum++ thing feels wrong...

have to instead look it up in the newDocNums?  perhaps by doing something like... newDocNums[segmentI][docNum]

Reply:
Hmm, let me think a little bit more about this..

Reply:
>> ... perhaps by doing something like... newDocNums[segmentI][docNum]
> Hmm, let me think a little bit more about this..

Basically, the idea is that it has to remap from an old docNum to the right new doc num.  Just doing a "newDocNum++" would be assigning an incorrect different number.

Reply:
Yes I agree on that, but I think newDocNums[segmentI][docNum] wouldn't fetch us the right newDocNum. Also I'm working through a unit test failure seen with this change, where certain terms that should've been present in certain old segments aren't..

Reply:
> ... I think newDocNums[segmentI][docNum] wouldn't fetch us the right newDocNum. Also I'm working through a unit test failure seen with this change, where certain terms that should've been present in certain old segments aren't.

good point -- looks like the "segmentI" lookup int has a different "basis" w.r.t. newDocNums vs the segmentsInFocus

Reply:
> good point -- looks like the "segmentI" lookup int has a different "basis" w.r.t. newDocNums vs the segmentsInFocus

uh, nevermind -- i might have misunderstood that (where I was incorrectly looking at or thinking about "newDocsNumIn" when I wrote that)

Reply:
oh, so segmentsInFocus is intended to keep the segmentI lookup correct. Its the second 'docNum' lookup which doesn't seem right.

Reply:
A slight correction on my last comment here - the unit test failure is because certain segments don't have an initialized fieldDvReader for the fieldID.